Creating Curb Objects

jacobmalleau opened the pull request:
## Explain pull request

First commit / defining Curb Object

## Is this a breaking change

* No, not breaking

## Impacted Spec

Which API(s) will this pull request impact?

* `Curbs`
* `Events`
* `Metrics`

## Additional context

Add any other context or screenshots about the feature request here.

jacobmalleau committed:
Update README.md

First commit / defining Curb Object

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/openmobilityfoundation/curb-data-specification?pullRequest=126) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/openmobilityfoundation/curb-data-specification?pullRequest=126) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/openmobilityfoundation/curb-data-specification?pullRequest=126) it.</sub>

jacobmalleau committed:
Update Curb Object Attributes

schnuerle committed:
Added object basic properties, well known types, TOC

jacobmalleau committed:
Add Objects to Zones and Spaces

jacobmalleau committed:
Added more and sorted object types

jacobmalleau committed:
Add objects to events

jacobmalleau committed:
Adding Objects to Metrics

jacobmalleau committed:
Address Some Comments from Working Group Meeting

- Note centroid of object
- Note positive and negative distances
- Start to populate type specific attributes

schnuerle committed:
Adding Point GeoJSON

schnuerle committed:
Add object owner, external id/URL

schnuerle committed:
Added object endpoints

schnuerle committed:
Added Object Type Attributes

schnuerle committed:
Add operator field & space/zone endpoint parameters

schnuerle committed:
Merged object attributes to types

schnuerle committed:
Rearranged required object fields

schnuerle committed:
Added object shape

schnuerle committed:
Removed duplicate geometry field

jacobmalleau committed:
Update README.md

Updated language around Object Type Attributes based on latest discussions.

It omits details on specific attributes for future versions of the spec.

jacobmalleau committed:
Added Objects to Event Endpoints

jacobmalleau committed:
Added Objects to Metric Endpoints

jacobmalleau committed:
Noted Objects are like Spaces, no previous IDs needed for updates

jamesian-barr commented:
@jacobmalleau this PR appears to be in decent shape. What do we need to do to take this to review?

schnuerle commented:
@jamesian-barr there's probably a few more things to finish and revise on here. I'll make some targeted comments.

Then the steering committee (SC) can bring it before the working group for review. There is a SC meeting tomorrow so we can see if this is ready to present.

schnuerle commented on curbs/README.md:
```diff
@@ -325,6 +372,96 @@ A Curb Space is represented as a JSON object whose fields are as follows:
+
+Some object types will have unique, custom attributes that can be optionally included with the given object type. These are given as unordered key-value pairs, e.g. `kilowatts` :  `1.21`. For this current version of Curb Objects, specific object type attributes will not be included in the specification. As use of Curb Objects grows or specific use cases become more involved, common object type at ...
+
+##### Sample Object Type Attributes (for Future)
```
This area needs to be fleshed out, and could get complicated fast. It's for custom object fields, so some subset of object types above would need to have their custom fields defined.

jacobmalleau replied:
I think we would be fine to leave this as a list of examples. 
Do we have to tie these attributes to specific object types? Or just leave it open for cities as they see fit.

schnuerle replied:
We agreed to remove this entire custom object type attributes section. Make sure references to it are removed from other parts of Objects.

jacobmalleau replied:
Added in latest commit. I also added this for reference

Unique object type attributes for each object type are not included in this version of the specification. As use of Curb Objects grows or specific use cases become more involved, common object type attributes can be added.

schnuerle replied:
People on the call agreed that leaving it in would create a breaking change later, and extend the timeline to a future major release (2.0) vs a minor release sooner (1.2). 

It would also create more work now and extend the timeline because for each type of object that already has some type of standard or defintion (EV chargers, trees, parking meters, bike racks, etc) we would have decide on which to use, what to include, and what to modify. 

I think we should leave it out and if people want to create a draft of a version of this they are using, then bring it back to the WG, it will be easier to add later (this is a typical way the spec gets updated).

jacobmalleau replied:
Agreed on leaving it out for now and we can see how Curb Objects is used to add in custom object type attributes later

alexdemisch replied:
@Mu-yi-Zhou @felixir @mschwartzie To achieve our goal of storing all of the asset data proposed in the SFMTA Internal Curb Data Model (shared by @Mu-yi-Zhou at a previous CDSWG meeting), one option is to do as @schnuerle proposes and implement the custom attribute key value pairs ourselves in our own fork, see how well it works, then come back to the working group.

schnuerle replied:
Alright based on conversations with SFMTA and INRIX and OMF, we will leave out custom object attributes for now, and get this PR pushed to the development branch for easier reference and use by all parties. As SFMTA works on implementation, they may bring back other base field suggestions to Objects, and/or a proposal for how to implement custom attributes. If this is done while we are still working on 1.1 in the coming months, we may be able to work it in. If it's done after the 1.1 release approval process is started or complete (likely early Q2 2025) then these ideas could be instead incorporated into a future CDS release.

schnuerle commented on events/README.md:
```diff
@@ -106,6 +108,7 @@ A Curb Event is represented as a JSON object, whose fields are as follows:
 | `curb_zone_id` | [UUID][uuid] | Conditionally Required | Unique ID of the Curb Zone where the event occurred. Required for events that occurred at a known Curb Zone for ALL _event_types_. |
 | `curb_area_ids` | Array of [UUID][uuid] | Conditionally Required | Unique IDs of the Curb Area where the event occurred. Since Curb Areas can overlap, an event may happen in more than one. Required for events that occurred in a known Curb Area, if known and used, for these event_types: _enter_area, exit_area, park_start, park_end_ |
 | `curb_space_id` | [UUID][uuid] | Conditionally Required | Unique ID of the Curb Space where the event occurred. Required for events that occurred at a known Curb Space, if known and used, for these event_types: _park_start, park_end, enter_area, exit_area_ |
+| `curb_object_id` | [UUID][uuid] | Conditionally Required | Unique ID of the Curb Object where the event occurred. Required for events that occurred at a known Curb Object, if known and used, for these event_types: _park_start, park_end, enter_area, exit_area_ |
```
Should this be for multiple IDs, so `curb_object_id`? One event could use a meter and an EV charger and a locker, for example.

jacobmalleau replied:
We could list multiple IDs here, although maybe it should only be what the true source of the event is (if you plug into the charger, book the locker, or pay at the meter)? I think leaving it as one option could be cleaner.

schnuerle replied:
Agreed that this can stay as one single ID per event. If multiple objects need to record simultaneous events at a zone, then they each get their own event. So this is fine as is.

schnuerle commented:
We had a great public working group meeting on this topic. [Here](https://github.com/openmobilityfoundation/curb-data-specification/wiki/Web-Conference-2024.10.29-Curb) are the action items, notes, and recording. 

Here were the main consensus items from the 49 attendees, including direct feedback from representatives from CurbIQ, Umojo, Turnstone, INRIX, SFTMA, SDOT, and others. 

1. Remove custom attributes from Curb Objects (could come later in 1.2 if it's needed)
2. Update curb object ids description in zones to specify it is area of influence
3. Update curb zone id description in Objects to specify it is location of the object at which zone 

I'll be flagging the specific PR lines with each of these for @jacobmalleau to edit. If you agree with these, let us know here. If don't agree, let us know too.

schnuerle commented on curbs/README.md:
```diff
@@ -263,6 +308,7 @@ A Curb Zone is represented as a JSON object, whose fields are as follows:
 | `entire_roadway`| Boolean | Optional | If "true", this curb location takes up the entire width of the roadway (which may be impassible for through traffic when the Curb Zone is being used for parking or loading). This is a common condition for alleyways. If `entire_roadway` is `true`, `street_side` MUST NOT be present. |
 | `curb_area_ids`| Array of [UUID][uuid] | Optional | The ID(s) of the [Curb Areas](#curb-area) that this Curb Zone is a part of. If specified, the areas identified MUST be retrievable through the Curb API and its geographical area MUST contain that of the Curb Zone. |
 | `curb_space_ids`| Array of [UUID][uuid] | Optional | The ID(s) of the [Curb Spaces](#curb-space) that this Curb Zone contains. If specified, the spaces identified MUST be retrievable through the Curb API and its geographical area MUST be contained in this Curb Zone. |
+| `curb_object_ids` | Array of [UUID][uuid] | Optional | The ID(s) of the [Curb Objects](#curb-object) that this Curb Zone is related to. If specified, the objects identified MUST be retrievable through the Curb API. Curb Objects can be related to a Curb Space or a Curb Zone. |
```
This description should specify that the curb objects listed are the objects that are in its 'area of influence'. The Objects that relate to this curb zone. So a locker's ID might be listed in 4 curb zones along a block face. Or a camera's ID might be listed in 3 curb zones that the camera has visibility into. Or the parking meter might be useable by 5 zones on a block.

jacobmalleau replied:
Added text for this in the latest commit. I am still a little weary of how this can be interpreted (e.g. a camera is located in one curb zone, but monitors several curb zones, how will a user know which is which). But I think with the lat/long available for an object this should be covered (e.g. you use the lat/long to specify where the curb object is, then associate it with any zones it covers). Another example: a pay station's lat/long shows where it is physically, and this field lets you indicate all the curb zones it covers.

schnuerle replied:
Resolving so we can review the latest language in a new comment

schnuerle commented on curbs/README.md:
```diff
@@ -317,6 +363,7 @@ A Curb Space is represented as a JSON object whose fields are as follows:
 | `published_date` | [Timestamp][ts] | Required | The date/time that this curb area was first published in this data feed. |
 | `last_updated_date` | [Timestamp][ts] | Required | The date/time that the properties of ths curb area were last updated. This helps consumers know that some fields may have changed. |
 | `curb_zone_id` | [UUID][uuid] | Required | The ID of the Curb Zone this space is within. The geometry of the specified Curb Zone MUST contain the geometry of this space. |
+| `curb_object_ids` | Array of [UUID][uuid] | Conditionally Required | The ID(s) of the [Curb Objects](#curb-object) that this Curb Space is related to. If specified, the objects identified MUST be retrievable through the Curb API. Curb Objects can be related to a Curb Space or a Curb Zone.|
```
Similar to the object IDs in a zone, for spaces, this description should specify that the curb objects listed are the objects that are in the space's 'area of influence'. So a meter might be for 2 curb spaces.

jacobmalleau replied:
Addressed in latest commit

schnuerle replied:
Resolving so we can review the latest language in a new comment

schnuerle commented on curbs/README.md:
```diff
@@ -325,6 +372,96 @@ A Curb Space is represented as a JSON object whose fields are as follows:
+| ------ | ------ | ------------------- | ------------- |
+| `curb_object_id` | [UUID][uuid] | Required | The ID of the curb object. |
+| `geometry` | [Point][point] | Required |The spatial location of this curb object location. This can represent the approximate center of the object, or the centroid location of the object, depending on its size and shape. |
+| `curb_zone_id` | [UUID][uuid] | Conditionally Required | The ID of the Curb Zone this object is related to. The geometry of the specified Curb Zone does not need to directly relate to the geometry of this object. Either a Zone or Space ID is required for an Object. |
```
The curb_zone_id description should reflect the idea that the object's physical location is nearest to this zone. Similarly with curb_space_id below.

jacobmalleau replied:
Added in latest commit

felixir replied:
It would be great to also be able to relate the object to a policy id for signs.

mschwartzie replied:
To @felixir's point, INRIX has a specific implementation underway that relates sign assets to policies

schnuerle replied:
Ah so adding a `policy_id` to the curb object defintion? I guess that makes some sense, if the object at a curb zone is a sign that has one description on it and that aligns to just one policy. But most curb zone policies are not defined by just one sign? I'm not sure it's needed to do this, since the zone is connected to a policy already, and this would apply just one type of object, `signage`. Could you give us some use cases and reasons why you would need the `policy_id` field in an Object?

jacobmalleau replied:
Yeah, I think we thought that you could relate the object to the policy through the curb zone. Similar to how events can be related to policies through the curb zone.

schnuerle replied:
See the new curb_policy_id  field @felixir @Mu-yi-Zhou @alexdemisch @mschwartzie and give feedback on this

jacobmalleau committed:
Updates to Curb Objects Based on October 29th CDS Meeting

See meeting notes for additional context.

jacobmalleau reviewed (commented):
Left comments throughout with items addressed in my latest commit

felixir commented on curbs/README.md:
```diff
@@ -325,6 +372,96 @@ A Curb Space is represented as a JSON object whose fields are as follows:
+| Name   | Type   | Required/Optional   | Description   |
+| ------ | ------ | ------------------- | ------------- |
+| `curb_object_id` | [UUID][uuid] | Required | The ID of the curb object. |
+| `geometry` | [Point][point] | Required |The spatial location of this curb object location. This can represent the approximate center of the object, or the centroid location of the object, depending on its size and shape. |
```
It would be great to have this be more flexible and allow line and polygon for objects like bike racks and bus stops. This would simplify the need for linear distance and max length.

schnuerle replied:
This point represents a required field. 

We also have an optional `object_shape` field below which is for polygons. 

I don't think lines work in the context of defining where objects area (though you could argue they do with curb zones, but that's not the scope of this).

Mu-yi-Zhou commented:
Is Object Type Enum or free-form? If Enum, will be good to keep custom attributes or even a comment/description fields. Example user case - some assets in our database are owned by other city department or private developer, we need a filed to note that.

schnuerle commented on curbs/README.md:
```diff
@@ -263,6 +308,7 @@ A Curb Zone is represented as a JSON object, whose fields are as follows:
 | `entire_roadway`| Boolean | Optional | If "true", this curb location takes up the entire width of the roadway (which may be impassible for through traffic when the Curb Zone is being used for parking or loading). This is a common condition for alleyways. If `entire_roadway` is `true`, `street_side` MUST NOT be present. |
 | `curb_area_ids`| Array of [UUID][uuid] | Optional | The ID(s) of the [Curb Areas](#curb-area) that this Curb Zone is a part of. If specified, the areas identified MUST be retrievable through the Curb API and its geographical area MUST contain that of the Curb Zone. |
 | `curb_space_ids`| Array of [UUID][uuid] | Optional | The ID(s) of the [Curb Spaces](#curb-space) that this Curb Zone contains. If specified, the spaces identified MUST be retrievable through the Curb API and its geographical area MUST be contained in this Curb Zone. |
+| `curb_object_ids` | Array of [UUID][uuid] | Optional | The ID(s) of the [Curb Objects](#curb-object) that this Curb Zone is related to, in particular what Objects are in the Zone's areas of influence. For example, a pay station being used for multiple paid parking zones, a locker for a commercial loading zone, or a camera monitoring several zones. If specified, the objects identified MUST be re ...
```
Flagging this update for review

schnuerle commented on curbs/README.md:
```diff
@@ -317,6 +363,7 @@ A Curb Space is represented as a JSON object whose fields are as follows:
 | `published_date` | [Timestamp][ts] | Required | The date/time that this curb area was first published in this data feed. |
 | `last_updated_date` | [Timestamp][ts] | Required | The date/time that the properties of ths curb area were last updated. This helps consumers know that some fields may have changed. |
 | `curb_zone_id` | [UUID][uuid] | Required | The ID of the Curb Zone this space is within. The geometry of the specified Curb Zone MUST contain the geometry of this space. |
+| `curb_object_ids` | Array of [UUID][uuid] | Conditionally Required | The ID(s) of the [Curb Objects](#curb-object) that this Curb Space is related to, in particular what Objects are in the Space's areas of influence. For example, a meter being used for two paid parking spcaes, a locker for a commercial loading space, or a camera monitoring several spaces. If specified, the objects identified MU ...
```
Flagging this update for review and comment.

schnuerle commented:
> Is Object Type Enum or free-form? If Enum, will be good to keep custom attributes or even a comment/description fields. Example user case - some assets in our database are owned by other city department or private developer, we need a filed to note that.

Hi @Mu-yi-Zhou - The Curb Object object has lots of these freeform fields now as you can see [here](https://github.com/openmobilityfoundation/curb-data-specification/blob/c8f5dc00e4f307ed5a1abf5376fceea7d4d3cddc/curbs/README.md#curb-object): name, description, owner, operator, external object id, external URL. I think these cover your use case.

schnuerle commented:
Hi everyone, please take a look at the latest unresolved comments and things for review and leave your thoughts.

schnuerle commented:
The only things open for final consensus are now:

1. Do we need an optional polyline for objects, and what is the use case for this?
2. Should we add a policy_id field to objects, or does its connection to Zones and therefore policy_ids that way sufficient?
3. Review of the updated language clarifying how curb_object_ids are defined

alexdemisch commented:
Re: (1), a polyline would be helpful to represent curb paint, which we are planning to store as an asset/curb object, and represent as a line feature.

Mu-yi-Zhou commented:
2) Should we add a policy_id field to objects, or does its connection to Zones and therefore policy_ids that way sufficient? Having policy_id in Curb Object is useful. Although we can link a object with curb_zone then with policy, a curb zone can be associated with several policies.

schnuerle committed:
Adding policy id and polyline options for Objects

schnuerle committed:
Added LineString geometry type

schnuerle committed:
Updated polyline to linestring for spec consistency

schnuerle committed:
Typo in curb_policy_id type

schnuerle commented on curbs/README.md:
```diff
@@ -325,6 +372,87 @@ A Curb Space is represented as a JSON object whose fields are as follows:
+| `geometry` | [Point][point] | Required |The spatial location of this curb object location. This can represent the approximate center of the object, or the centroid location of the object, depending on its size and shape. |
+| `curb_zone_id` | [UUID][uuid] | Conditionally Required | The ID of the Curb Zone this object is physically in or closest to. The geometry of the specified Curb Zone does not need to directly relate to the geometry of this object. Either a Zone or Space ID is required for an Object. |
+| `curb_space_id` | [UUID][uuid] | Conditionally Required | The ID of the Curb Space this object is physically in or closest to. The geometry of the specified Curb Space does not need to directly relate to the geometry of this object. Either a Zone or Space ID is required for an Object. |
+| `curb_policy_id` | [UUID][uuid] | Optional | ID of [Policy object](#policy) that is directly associcated with this curb object. For example, `signage` or `paint` that relates to a single policy. |
```
See the new `curb_policy_id ` field @felixir @Mu-yi-Zhou @alexdemisch and give feedback on this

schnuerle committed:
Updated the curb object <-> policy relationship description

bhamlinSDOT commented:
> Re: (1), a polyline would be helpful to represent curb paint, which we are planning to store as an asset/curb object, and represent as a line feature.

@alexdemisch Can you explain more of the use cases for why you want curb paint represented in CDS? SDOT also has [several curb paint colors associated with our regulations](https://www.seattle.gov/transportation/projects-and-programs/programs/parking-program/parking-regulations/curb-colors) but I've never thought about listing the colors in CDS until I read your comment. Seems like important to have noted in asset mgmt. system but feels different than say a sign object listed in CDS. Curious your thoughts.

alexdemisch commented:
@bhamlinSDOT We plan to link regulations to the assets that define them to help maintain the data.  In some cases, we only define regulations through curb paint, so having a line feature would help us do that.  

Your comment just made me realize that we don't have anywhere to store the actual color of the curb paint in CDS.  I suppose that's fine for now as we'll have the linkage to the internal asset_id via the `external_object_id` field.  If we decide to expand on object attributes, it could go there.  Admittedly, the asset/regulation linkage might only be useful for internal purposes, but not sure.

jacobmalleau commented:
The main issue I see with adding a polyline option is that your Curb Objects dataset would then have both point and linear data which would make the data not compatible in GIS software unless separated out. 
A work around would be to have an additional polyline attribute to store the line geometry, but still having a point represent the object.

schnuerle commented:
> The main issue I see with adding a polyline option is that your Curb Objects dataset would then have both point and linear data which would make the data not compatible in GIS software unless separated out. A work around would be to have an additional polyline attribute to store the line geometry, but still having a point represent the object.

@jacobmalleau It seems like this PR supports your workaround idea. Can you explain why it's a workaround in your estimation? It seems as written there is a required point for every object, and there is an optional polyline field.  So it is separated out. How does that create compatibility issues with GIS software?

schnuerle commented:
Pushed to `dev` branch as is. Please continue the conversation here.

If needed we can make a new Issue or PR now to address suggested changes to `dev` before the 1.1 release starts the OMF approval process, at a later date.
